### DISCUSSION TITLE
Run syntax checking at save time rather than continuously

### DISCUSSION THREAD

**Discussion by bguil** [ocamllabs/vscode-ocaml-platform#1280] [Q&A]
When writing code, there are many times when the code is not correct until some future point when a function is finished.
For example, when someone writes pattern-matching cases, they know that the pattern-matching is not exhaustive and the highlighting of a large part of the code is annoying.

Is there a way to have the control of syntax, typing, exhaustivity of pattern matching only at save time?

**Comment by yawaramin** [comment]
Related to this, is there a way to turn off the pop-up message `ocamlformat: ignoring "..."...Syntax Error`? I'm in the middle of editing and I have syntax checks already highlighted in the file. I don't need a separate pop-up that keeps appearing periodically while I'm editing and distracting me. It breaks me out of my flow and I have to train myself to just ignore it...

**Reply by voodoos** [reply]
I guess we could filter out these messages and not send the "showMessage" notification here:

https://github.com/ocaml/ocaml-lsp/blob/671fbedc2bbe51b93f9a78d4344cbfa82a109d3f/ocaml-lsp-server/src/ocaml_lsp_server.ml#L329

**Comment by voodoos** [comment]
> Is there a way to have the control of syntax, typing, exhaustivity of pattern matching only at save time?

I think the LSP client in VScode does not allow this kind of configuration, which is very sad. Overall the client and vscode's API are very rigid and prevent most customization. Another idea would be to make the error decorations less invasive when they cover multiple lines, I don't know if that is possible or not.

However, in most cases, Merlin should be able to recover from an incorrectly file to propose meaningful suggestion and prevent cascading errors, but that's often not working as expected right now. Opening issues on Merlin with simple replications for cases that are not satisfying would be very useful to start fixing these. (this won't help with non-exhaustive pattern matching however...)
